util/failover.go

```diff
@@ -316,3 +309,85 @@ func GetPreferredNodes(clientset *kubernetes.Clientset, selector, namespace stri
 
 	return nodes, err
 }
+
+// ToggleAutoFailover enables or disables autofailover for a cluster.  Disabling autofailover means "pausing"
+// Patroni, which will result in Patroni stepping aside from managing the cluster.  This will effectively cause
+// Patroni to stop responding to failures or other database activities, e.g. it will not attempt to start the
+// database when stopped to perform maintenance
+func ToggleAutoFailover(clientset *kubernetes.Clientset, enable bool, pghaScope, namespace string) error  {
+	
+	// find the "config" configMap created by Patroni
+	configMapName := pghaScope+"-config"
+	log.Debugf("setting autofailover to %t for cluster with pgha scope %s", enable, pghaScope)
+
+	configMap, found := kubeapi.GetConfigMap(clientset, configMapName, namespace)
+	if !found {
+		err := fmt.Errorf("Unable to find configMap %s when attempting disable autofailover", configMapName)
+		log.Error(err)
+		return err
+	}
+
+	configJSONStr := configMap.ObjectMeta.Annotations["config"]
+
+	var configJSON map[string]interface{}
+	json.Unmarshal([]byte(configJSONStr), &configJSON)
+
+	if !enable {  
+		// disable autofail condition
+		disableFailover(clientset, configMap, configJSON, namespace)
+	}  else {
+		// enable autofail
+		enableFailover(clientset, configMap, configJSON, namespace)
+	}
+
+	return nil
+}
+
+// If "pause" is present in the config and set to "true", then it needs to be removed to enable failver.  
+// Otherwise, if "pause" isn't present in the config or if it has a value other than true, then assume autofail 
+// is enabled and do nothing (when Patroni see's an invalid value for "pause" it sets it to "true")
+func enableFailover(clientset *kubernetes.Clientset, configMap *v1.ConfigMap, configJSON map[string]interface{}, 
+	namespace string) error {
+	if _,ok := configJSON["pause"] ; ok || configJSON["pause"] == true  {
+		log.Debugf("updating pause key in configMap %s to enable autofailover", configMap.Name)
+		//  disabled autofail by removing "pause" from the config
+		delete(configJSON, "pause")
+		configJSONFinalStr, err := json.Marshal(configJSON)
+		if err != nil {
+			return err
+		}
+		configMap.ObjectMeta.Annotations["config"] = string(configJSONFinalStr)
+		err = kubeapi.UpdateConfigMap(clientset, configMap, namespace)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Debugf("autofailover already enabled according to the pause key (or lack thereof) in configMap %s", 
+			configMap.Name)
+	}
+	return nil
+}
+
+// If "pause" isn't present in the config then assume autofail is enabled and needs to be disabled by setting "pause
+// to true.  Or if it is present and set to "false", then it also needs to be disabled by setting "pause" to true.
+func disableFailover(clientset *kubernetes.Clientset, configMap *v1.ConfigMap, configJSON map[string]interface{}, 
+	namespace string) error {
+	if _,ok := configJSON["pause"] ; !ok || configJSON["pause"] == false  {
```

Comment:
Same thoughts here as above.

Reply:
I think the logic holds here, i.e. I want to take action if the `pause` key doesn't exist in the map at all, or if it exists but is set to `false`.

Reply:
I see what you're getting at and don't necessarily disagree. Though, candidly, it still feels grotty to me.

Two reasons:

1) `configJSON` is a `map[string]interface{}`. What happens if the value represented by the key "pause" is ever anything other than a `bool`? Unless you assert (safely) you have no protection for that case. Also a key benefit of the the proposed solution is that you get that safety along with the safe assumption that if it is missing it is assumed to be false.
2) Kind of ties into the above... `value == bool` vs `value` or `!value`. In the first, you are making assumptions about `value` which can be unsafe to do given the nature of the map being `map[string]interface{}`. It would be better to operate on something you actually KNOW to be a `bool`. And unfortunately, that's not guaranteed without the assertion, which should be done safely.

I still think something like the following would be cleaner and 'safer' in the long run.

```
if value, ok := config["pause"].(bool); !ok || !value { ... }
```

Just my two cents. :-)

Reply:
Appreciate the input here, this one got me thinking.  The one things that "bugs" me with this approach (and that is admittedly not currently relevant to this function), is that I loose visibility into whether not the key even existed in the map in the first place, or if it was simply set to an invalid value (i.e. both the key missing altogether and an invalid key [e.g. some random string] result in `ok=false`).

Also, I'm not clearly seeing what protection I'm missing by using `value == bool`.  Specifically, I know in this case that I only care that the result is (or with my recent update _isn't_) a certain boolean value, i.e. that it isn't `true` in the following case (and I know that Go will store a `bool` for a valid JSON boolean, which is what I want):

`configJSON["pause"] != true `

...and therefore I can store any other value of any type (null, string, int, etc.) in the map for the "pause" key and the result will be the same, i.e. the condition will fail.

Reply:
I think the current construct `_,ok := configJSON["pause"] ; ok && configJSON["pause"] == true` is certainly clearer and achieves what Adam recommended via a quirk of the language. (It seems to perform either an implicit type conversion or immediately understands that a value of type `interface{}` is not equal to `bool`(true))
However, there's a behavioral caution here that wasn't captured in Adam's recommendation and that's concerning the underlying type of the JSON. In deserializing the JSON, we're going from a weakly typed system to a strongly typed one, which presents opportunities for this type of failure:
https://play.golang.com/p/v1Xr42-qA51
That is, the check will fail if the value is not of `bool` type according to javascript. In this case, using `ParseBool` on the value would ensure correct behavior without someone having to get the type correct in the JSON. (Think about someone modifying the CRD directly and including quotes when they shouldn't have)

Reply:
FWIW, I'm fine either way on deciding to postpone the failure scenario, but do recognize it's leaving a potential trap.

Reply:
> Appreciate the input here, this one got me thinking. The one things that "bugs" me with this approach (and that is admittedly not currently relevant to this function), is that I loose visibility into whether not the key even existed in the map in the first place, or if it was simply set to an invalid value (i.e. both the key missing altogether and an invalid key [e.g. some random string] result in ok=false).

Sure, but that's a different case altogether though, right? The point is that as soon as you want to operate on the value represented by the key, you want to know it's type. Otherwise, you're at risk of a `panic`, which I do think is worth protecting against.

> Also, I'm not clearly seeing what protection I'm missing by using value == bool. Specifically, I know in this case that I only care that the result is (or with my recent update isn't) a certain boolean value, i.e. that it isn't true in the following case (and I know that Go will store a bool for a valid JSON boolean, which is what I want):

Well, there are two things here:

1) `if bool == bool { ... }` is really just `if bool { ... }` right? perhaps that's just a stylistic preference, but doing an extra 'comparison' to determine the value of what you (should) already have seems unnecessary to me.

2) There is value in actually knowing and guaranteeing the type of the thing you are comparing. Making assumptions about the type, in my view, is inherently dangerous.

Essentially, interface values are considered equal when their underlying concrete types and values are comparable and equal. However, if the underlying concrete types are not comparable any attempt to compare them will cause a runtime panic.

> ...and therefore I can store any other value of any type (null, string, int, etc.) in the map for the "pause" key and the result will be the same, i.e. the condition will fail.

I see your point, but I feel like this is an misuse of `interface{}`. It's nice to have the 'flexibility' but when operating on the underlying type, it's best, in my view, to ensure you're actually operating on that concrete type.

Ultimately, for me, it's about type safety and safe guarding against panics.  Maybe that case may never arise, but to me it still seems worth taking measures to guard against it.

Reply:
What value for `pause` would result in a panic?

https://play.golang.org/p/iz8pimpcO0G

Reply:
> What value for `pause` would result in a panic?

My read of this feedback is to be careful about using / relying on a value of type `interface{}` to "behave well" in all cases. In this _specific_ scenario, performing a comparison of a type `interface{}` against a literal, built-in type is guaranteed to be safe, but the pattern is not universally so.

Reply:
I think overall this is a very interesting discussion and certainly makes mention of some things we want to incorporate in the future especially around type safety, which we should certainly incorporate as the codebase evolves. In fact, this discussion could serve as a good case study and be applied to "design patterns" we want to incorporate into our go code.

However, I am satisfied right now with what we have and so long as the code is correct, I advise we move on.